File: tools/device-share-hub/ios/Album/RemoteRelayPresence.swift
import Foundation

/// Re-authenticates and publishes relay presence only after a profile was enrolled.
final class RemoteRelayPresence {
    private let queue = DispatchQueue(label: "com.zwm.album.remote-presence")
    private var timer: DispatchSourceTimer?
    private var session: RemoteRelayClient.Session?
    private var profileKey = ""
    private var retryAfter: TimeInterval = 0
    private var failureDelay: TimeInterval = 0

    func start() {
        queue.async { [weak self] in
            guard let self = self, self.timer == nil else { return }
            let timer = DispatchSource.makeTimerSource(queue: self.queue)
            timer.schedule(deadline: .now(), repeating: .seconds(10), leeway: .seconds(2))
            timer.setEventHandler { [weak self] in self?.tick() }
            timer.resume()
            self.timer = timer
        }
    }

    func stop() {
        queue.async { [weak self] in
            self?.timer?.cancel()
            self?.timer = nil
            self?.session = nil
        }
    }

    private func tick() {
        guard let profile = RemoteRelayProfile.load() else {
            session = nil
            profileKey = ""
            retryAfter = 0
            failureDelay = 0
            return
        }
        let nextProfileKey = "\(profile.endpoint)\n\(profile.certificateSignature)\n"
            + (profile.certificate["deviceId"] as? String ?? "")
        if nextProfileKey != profileKey {
            profileKey = nextProfileKey
            session = nil
            retryAfter = 0
            failureDelay = 0
        }
        let now = Date().timeIntervalSince1970
        if now < retryAfter { return }
        do {
            var current = session
            if current == nil || current!.expired
                || current!.endpoint != profile.endpoint
                || current!.deviceId != (profile.certificate["deviceId"] as? String ?? "") {
                current = try RemoteRelayClient.createSession(
                    endpoint: profile.endpoint,
                    certificate: profile.certificate,
                    certificateSignature: profile.certificateSignature
                )
                session = current
            }
            if let current = current { try RemoteRelayClient.heartbeat(current) }
            retryAfter = 0
            failureDelay = 0
        } catch {
            session = nil
            failureDelay = min(5 * 60, failureDelay <= 0 ? 10 : failureDelay * 2)
            retryAfter = now + failureDelay
        }
    }
}

File: tools/device-share-hub/android/app/src/main/java/com/zwm/gallery/RemoteRelayPresence.java
package com.zwm.gallery;

import android.content.Context;

import org.json.JSONArray;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/** Re-authenticates and publishes relay presence only after a profile was enrolled. */
final class RemoteRelayPresence {
    interface Listener {
        void onInbox(RemoteRelayClient.Session session, JSONArray transfers);

        default void onP2PSessions(RemoteRelayClient.Session session, JSONArray sessions) { }
    }

    private final Context context;
    private final Listener listener;
    private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
    private volatile RemoteRelayClient.Session session;
    private volatile boolean started;
    private String profileKey = "";
    private long retryAfterMs;
    private long failureDelayMs;

    RemoteRelayPresence(Context context, Listener listener) {
        this.context = context.getApplicationContext();
        this.listener = listener;
    }

    void start() {
        if (started) return;
        started = true;
        executor.scheduleWithFixedDelay(this::tick, 0, 10, TimeUnit.SECONDS);
    }

    void stop() {
        started = false;
        session = null;
        executor.shutdownNow();
    }

    private void tick() {
        if (!started) return;
        RemoteRelayProfile.Profile profile = RemoteRelayProfile.load(context);
        if (profile == null) {
            session = null;
            profileKey = "";
            retryAfterMs = 0;
            failureDelayMs = 0;
            return;
        }
        String nextProfileKey = profile.endpoint + "\n" + profile.certificateSignature
                + "\n" + profile.certificate.optString("deviceId", "");
        if (!nextProfileKey.equals(profileKey)) {
            profileKey = nextProfileKey;
            session = null;
            retryAfterMs = 0;
            failureDelayMs = 0;
        }
        long now = System.currentTimeMillis();
        if (now < retryAfterMs) return;
        try {
            RemoteRelayClient.Session current = session;
            if (current == null || current.expired(System.currentTimeMillis())
                    || !current.endpoint.equals(profile.endpoint)
                    || !current.deviceId.equals(profile.certificate.getString("deviceId"))) {
                current = RemoteRelayClient.createSession(context, profile.endpoint,
                        profile.certificate, profile.certificateSignature);
                session = current;
            }
            RemoteRelayClient.heartbeat(current);
            JSONArray transfers = RemoteRelayClient.inbox(current);
            JSONArray p2pSessions = RemoteRelayClient.p2pSessions(current);
            if (listener != null) {
                try {
                    listener.onInbox(current, transfers);
                    listener.onP2PSessions(current, p2pSessions);
                } catch (RuntimeException error) {
                    DiagnosticLog.write(context, "remote_inbox_listener_failed",
                            error.getClass().getSimpleName());
                }
            }
            retryAfterMs = 0;
            failureDelayMs = 0;
        } catch (Exception error) {
            session = null;
            long delay = failureDelayMs <= 0 ? 10_000L : Math.min(5 * 60_000L, failureDelayMs * 2);
            failureDelayMs = delay;
            retryAfterMs = now + delay;
            DiagnosticLog.write(context, "remote_presence_failed", error.getClass().getSimpleName());
        }
    }
}
